cityofaustin/atd-vz-data
# LDM cr3 crash location trigger

## Associated issues
Closes https://github.com/cityofaustin/atd-data-tech/issues/12646

So originally I thought the `cris_facts.atd_txdot_crashes` table shouldn't have a `location_id` column since that's not a value we get from CRIS but that proved to be a reallyyyyy difficult and complicated pattern to carry out in these triggers. It involved a lot of extra steps and complex logic. In the end I'm deciding this solution is best, to have the `location_id` column in both schemas and in the combined view it will either be the `location_id` from the `cris_facts` or `vz_facts` if there is a non-null value for it there. Now with some thought this solution makes more sense to me in general, it would have been weird imo for there to be `location_id` in the `vz_facts.atd_txdot_crashes` that really represented the location of the coordinates in the `cris_facts` schema.

A follow up to this issue is to address the stuff in the TODO section, the UI now needs to match up to these changes and stop using the `latitude_primary` and `longitude_primary` columns.

## Testing
**URL to test:** 
Local

**Steps to test:**

1. Pull this branch locally.
2. Replicate from prod locally by running `./vision-zero -l replicate-db`
3. From the atd-vzd folder, run `hasura migrate apply --envfile .env.local` and then `hasura metadata apply --envfile .env.local`
4. Test inserts to `cris_facts.atd_txdot_crashes`:
```plpgsql
-- insert a cr3 crash with coordinates that will associate to location_id 8337406736
INSERT INTO cris_facts.atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.16841600, -97.78449000, 9999);

-- insert a cr3 crash with coordinates that will not associate to a location
INSERT INTO cris_facts.atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (30.386136, -97.67309, 8888);

-- insert a cr3 crash without coordinates that will not associate to a location
INSERT INTO cris_facts.atd_txdot_crashes(latitude, longitude, crash_id)
VALUES (NULL, NULL, 7777);
```
5. Update an existing record with a `location_id` of NULL to have the `latitude` and `longitude` of a record with a location. After update, you should see the record associate with the same location as the record you took the lat/lon from.
6. Update an existing record with a `location_id` that is not null and update it with `latitude` and `longitude` coordinates from an existing record that has latitude and longitude and no location_id associated.
7. Now test editing records in `vz_facts.atd_txdot_crashes`.
```plpgsql
-- edit one of the records in vz_facts to associate to location_id 5B5D5E72E4
UPDATE vz_facts.atd_txdot_crashes
SET latitude = 30.298207784614654,
    longitude = -97.74852940415873
WHERE crash_id = 7777;
```
8. Check in the unified view `public.atd_txdot_crashes` that crash_id 7777 has the location_id from the vz_facts schema, 5B5D5E72E4.

---

### TODO:

- [ ] I think we should only have `latitude` and `longitude` columns, no more `latitude_primary` or `latitude_geocoded`. The `latitude_primary`/`longitude_primary` columns were used to keep track of lats from CRIS vs lats from VZ edits. We need to deprecate these other columns and adjust the code so that when users update coordinates through the UI map feature, they update the `latitude` `longitude` columns instead of the primary columns
- [ ] I also think we should deprecate the `position` column because once again it is repetitive data that opens up room for de normalization.


---
#### Ship list
